app/helpers/ExerciseConfig/Pipeline/Box/Boxes/PrologCompilationBox.php

```diff
@@ -0,0 +1,134 @@
+use Exception;
+
+
+/**
+ * Box which represents compilation of given prolog scripts.
+ */
+class PrologCompilationBox extends CompilationBox
+{
+  /** Type key */
+  public static $BOX_TYPE = "prolog-compilation";
+  public static $PROLOG_BINARY = "/usr/bin/swipl";
+  public static $DEFAULT_NAME = "SWI-Prolog Compilation";
+  public static $WRAPPER_FILE_PORT_KEY = "compilation-wrapper";
+
+  private static $initialized = false;
+  private static $defaultInputPorts;
+  private static $defaultOutputPorts;
+
+  /**
+   * Static initializer.
+   * @throws ExerciseConfigException
+   */
+  public static function init() {
+    if (!self::$initialized) {
+      self::$initialized = true;
+      self::$defaultInputPorts = array(
+        new Port((new PortMeta())->setName(self::$WRAPPER_FILE_PORT_KEY)->setType(VariableTypes::$FILE_TYPE)),
+        new Port((new PortMeta())->setName(self::$RUNNER_FILE_PORT_KEY)->setType(VariableTypes::$FILE_TYPE)),
+        new Port((new PortMeta())->setName(self::$ARGS_PORT_KEY)->setType(VariableTypes::$STRING_ARRAY_TYPE)),
+        new Port((new PortMeta())->setName(self::$SOURCE_FILES_PORT_KEY)->setType(VariableTypes::$FILE_ARRAY_TYPE)),
+        new Port((new PortMeta())->setName(self::$EXTRA_FILES_PORT_KEY)->setType(VariableTypes::$FILE_ARRAY_TYPE))
+      );
+      self::$defaultOutputPorts = array(
+        new Port((new PortMeta())->setName(self::$BINARY_FILE_PORT_KEY)->setType(VariableTypes::$FILE_TYPE))
+      );
+    }
+  }
+
+  /**
+   * JudgeNormalBox constructor.
+   * @param BoxMeta $meta
+   */
+  public function __construct(BoxMeta $meta) {
+    parent::__construct($meta);
+  }
+
+
+  /**
+   * Get type of this box.
+   * @return string
+   */
+  public function getType(): string {
+    return self::$BOX_TYPE;
+  }
+
+  /**
+   * Get default input ports for this box.
+   * @return array
+   * @throws ExerciseConfigException
+   */
+  public function getDefaultInputPorts(): array {
+    self::init();
+    return self::$defaultInputPorts;
+  }
+
+  /**
+   * Get default output ports for this box.
+   * @return array
+   * @throws ExerciseConfigException
+   */
+  public function getDefaultOutputPorts(): array {
+    self::init();
+    return self::$defaultOutputPorts;
+  }
+
+  /**
+   * Get default name of this box.
+   * @return string
+   */
+  public function getDefaultName(): string {
+    return self::$DEFAULT_NAME;
+  }
+
+
+  /**
+   * Compile box into set of low-level tasks.
+   * @param CompilationParams $params
+   * @return array
+   * @throws Exception
+   */
+  public function compile(CompilationParams $params): array {
+    $task = $this->compileBaseTask($params);
+
+    // prolog has special wrapper runner
+    $wrapper = $this->getInputPortValue(self::$WRAPPER_FILE_PORT_KEY)->getValue();
+    $task->setCommandBinary($wrapper);
+
+    $args = [ self::$PROLOG_BINARY ];
+    if ($this->hasInputPortValue(self::$ARGS_PORT_KEY)) {
+      $args = array_merge($args,
```

Comment:
Maybe it's just me, but I would use array_push to assemble $args. Less coding + less memory copying + more readable (in my opinion). Of course it will get the best results in combination with splat operator...